### PR TITLE
M3 lead bundle refactoring

### PR DESCRIPTION
**Pull request by escopecz**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/8000
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Refactoring of the LeadBundle needed to upgrade Symfony

#### Steps to test this PR:
- during testing focus on select boxes that the keys and values are not flipped.
1. Load up [this PR](https://mautibox.com)
1.a. If you are testing on your local, run `app/console doctrine:migrations:migrate`
2. Test contact forms. Creating, updating, importing contacts. Test all field types. Booleans, selects, multiselects, tags and so on.
3. Test segment forms
4. Test company forms
5. Test custom field forms

#### List deprecations along with the new alternative:
See https://docs.google.com/document/d/1lHNK8hHQtvyWlz9OtfAoLupDVb5-QunQjq-HAGlAsS4/edit#

#### List backwards compatibility breaks:
See https://docs.google.com/document/d/1lHNK8hHQtvyWlz9OtfAoLupDVb5-QunQjq-HAGlAsS4/edit#

### CODE REVIEW

**Comment by ondrejsibl** [comment]
@escopecz While editing single or multiple companies in contact, I get this error:
`/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:115 at
/vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:128 at Doctrine\DBAL\Driver\AbstractMySQLDriver -> convertException ( 'An exception occurred while executing 'INSERT INTO mautic_companies_leads (date_added, is_primary, company_id, lead_id) VALUES (?, ?, ?, ?)' with params ["2019-11-26 00:22:31", 1, 2, 2401]: SQLSTATE[HY000]: General error: 1364 Field 'manually_removed' doesn't have a default value', object(PDOException) )
/vendor/doctrine/dbal/lib/Doctrine/DBAL/Statement.php:177 at Doctrine\DBAL\DBALException :: driverExceptionDuringQuery ( object(Driver), object(PDOException), 'INSERT INTO mautic_companies_leads (date_added, is_primary, company_id, lead_id) VALUES (?, ?, ?, ?)', array('2019-11-26 00:22:31', '1', '2', '2401') )
/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:281 at Doctrine\DBAL\Statement -> execute ( )
/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:1014 at Doctrine\ORM\Persisters\Entity\BasicEntityPersister -> executeInserts ( )
/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:378 at Doctrine\ORM\UnitOfWork -> executeInserts ( object(ClassMetadata) )
/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:356 at Doctrine\ORM\UnitOfWork -> commit ( null )
/app/bundles/CoreBundle/Entity/CommonRepository.php:785 at Doctrine\ORM\EntityManager -> flush ( )
/app/bundles/LeadBundle/Entity/CompanyLeadRepository.php:42 at Mautic\CoreBundle\Entity\CommonRepository -> saveEntities ( array(object(CompanyLead), object(CompanyLead)) )
/app/bundles/LeadBundle/Model/CompanyModel.php:383 at Mautic\LeadBundle\Entity\CompanyLeadRepository -> saveEntities ( array(object(CompanyLead), object(CompanyLead)) )
/app/bundles/LeadBundle/Model/LeadModel.php:1878 at Mautic\LeadBundle\Model\CompanyModel -> addLeadToCompany ( array('2', '1'), object(Lead) )
/app/bundles/LeadBundle/Controller/LeadController.php:644 at Mautic\LeadBundle\Model\LeadModel -> modifyCompanies ( object(Lead), array('2', '1') )
/app/bundles/CoreBundle/Controller/CommonController.php:474 at Mautic\LeadBundle\Controller\LeadController -> editAction ( '2401', '' )
/vendor/symfony/http-kernel/HttpKernel.php:135 at Mautic\CoreBundle\Controller\CommonController -> executeAction ( 'edit', '2401', '0', '' )
/vendor/symfony/http-kernel/HttpKernel.php:57 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), '1' )
/vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php:67 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), '1', true )
/vendor/symfony/http-kernel/Kernel.php:183 at Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel -> handle ( object(Request), '1', true )
/app/AppKernel.php:141 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), '1', true )
/app/middlewares/CORSMiddleware.php:93 at AppKernel -> handle ( object(Request), '1', true )
/app/middlewares/CatchExceptionMiddleware.php:45 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/Dev/IpRestrictMiddleware.php:66 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/VersionCheckMiddleware.php:57 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), '1', true )
/vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), '1', true )
/vendor/stack/run/src/Stack/run.php:13 at Stack\StackedHttpKernel -> handle ( object(Request) )
/index_dev.php:33 at Stack\run ( object(StackedHttpKernel) )`

**Comment by ondrejsibl** [comment]
Missing translation while creating a new contact:
Snímek obrazovky 2019-11-26 v 1 36 13

**Comment by escopecz** [comment]
Thanks for testing @ondrejsibl!

1. I added new migration and updated steps to test.
2. I checked that the translation string exists and I can see it translated on my local. Don't know what the problem is. Do you still see that even after you clear cache?

**Comment by ondrejsibl** [comment]
@escopecz 1. Now it works! 2. Even if I clear cache, I can still see it. Yes, the translation string exists.

**Comment by ondrejsibl** [comment]
In segments, I see a problem with URL from email select field. At first, there are values yes and no. If I save it, it says that the value is invalid and it changes to 0 and 1.
Snímek obrazovky 2019-11-26 v 12 48 58

**Comment by ondrejsibl** [comment]
If I try to import contacts, I get error:
`Symfony\Component\Routing\Exception\RouteNotFoundException: Unable to generate a URL for the named route "mautic_contact_import_index" as such route does not exist.
/app/cache/dev/appDevDebugProjectContainerUrlGenerator.php:232
/app/cache/dev/appDevDebugProjectContainerUrlGenerator.php:232 at
/vendor/symfony/routing/Router.php:240 at appDevDebugProjectContainerUrlGenerator -> generate ( 'mautic_contact_import_index', array(), '1' )
/vendor/symfony/framework-bundle/Templating/Helper/RouterHelper.php:62 at Symfony\Component\Routing\Router -> generate ( 'mautic_contact_import_index', array(), '1' )
/app/bundles/CoreBundle/Views/Helper/page_actions.html.php:51 at Symfony\Bundle\FrameworkBundle\Templating\Helper\RouterHelper -> path ( 'mautic_contact_import_index', array() )
/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php:175 at require ( '/Users/ondrejsibl/dev/apps/community/app/bundles/CoreBundle/Views/Helper/page_actions.html.php' )
/vendor/symfony/templating/PhpEngine.php:78 at Mautic\CoreBundle\Templating\Engine\PhpEngine -> evaluate ( null, array('app' => object(GlobalVariables), 'routeBase' => 'contact_import', 'langVar' => 'lead.import', 'templateButtons' => array('close' => true), 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticCoreBundle:Helper:page_actions.html.php', 'view' => object(PhpEngine)) )
/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php:144 at Symfony\Component\Templating\PhpEngine -> render ( 'MauticCoreBundle:Helper:page_actions.html.php', array('app' => object(GlobalVariables), 'routeBase' => 'contact_import', 'langVar' => 'lead.import', 'templateButtons' => array('close' => true), 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticCoreBundle:Helper:page_actions.html.php') )
/app/bundles/LeadBundle/Views/Import/upload_form.html.php:17 at Mautic\CoreBundle\Templating\Engine\PhpEngine -> render ( 'MauticCoreBundle:Helper:page_actions.html.php', array('routeBase' => 'contact_import', 'langVar' => 'lead.import', 'templateButtons' => array('close' => true), 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticCoreBundle:Helper:page_actions.html.php') )
/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php:175 at require ( '/Users/ondrejsibl/dev/apps/community/app/bundles/LeadBundle/Views/Import/upload_form.html.php' )
/vendor/symfony/templating/PhpEngine.php:78 at Mautic\CoreBundle\Templating\Engine\PhpEngine -> evaluate ( null, array('app' => object(GlobalVariables), 'form' => object(FormView), 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticLeadBundle:Import:upload_form.html.php', 'view' => object(PhpEngine)) )
/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php:144 at Symfony\Component\Templating\PhpEngine -> render ( 'MauticLeadBundle:Import:upload_form.html.php', array('app' => object(GlobalVariables), 'form' => object(FormView), 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticLeadBundle:Import:upload_form.html.php') )
/app/bundles/LeadBundle/Views/Import/new.html.php:18 at Mautic\CoreBundle\Templating\Engine\PhpEngine -> render ( 'MauticLeadBundle:Import:upload_form.html.php', array('form' => object(FormView), 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticLeadBundle:Import:upload_form.html.php') )
/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php:175 at require ( '/Users/ondrejsibl/dev/apps/community/app/bundles/LeadBundle/Views/Import/new.html.php' )
/vendor/symfony/templating/PhpEngine.php:78 at Mautic\CoreBundle\Templating\Engine\PhpEngine -> evaluate ( null, array('app' => object(GlobalVariables), 'form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticLeadBundle:Import:new.html.php', 'view' => object(PhpEngine)) )
/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php:144 at Symfony\Component\Templating\PhpEngine -> render ( 'MauticLeadBundle:Import:new.html.php', array('app' => object(GlobalVariables), 'form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticLeadBundle:Import:new.html.php') )
/vendor/symfony/templating/DelegatingEngine.php:41 at Mautic\CoreBundle\Templating\Engine\PhpEngine -> render ( 'MauticLeadBundle:Import:new.html.php', array('form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport', 'mauticTemplate' => 'MauticLeadBundle:Import:new.html.php') )
/vendor/symfony/framework-bundle/Controller/Controller.php:163 at Symfony\Component\Templating\DelegatingEngine -> render ( 'MauticLeadBundle:Import:new.html.php', array('form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport') )
/app/bundles/CoreBundle/Controller/CommonController.php:393 at Symfony\Bundle\FrameworkBundle\Controller\Controller -> renderView ( 'MauticLeadBundle:Import:new.html.php', array('form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport') )
/app/bundles/CoreBundle/Controller/CommonController.php:224 at Mautic\CoreBundle\Controller\CommonController -> ajaxAction ( array('viewParameters' => array('form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport'), 'contentTemplate' => 'MauticLeadBundle:Import:new.html.php', 'passthroughVars' => array('activeLink' => '#mautic_contact_index', 'mauticContent' => 'leadImport', 'route' => '/index_dev.php/s/contacts/import/new', 'step' => '1', 'progress' => object(Progress))) )
/app/bundles/LeadBundle/Controller/ImportController.php:485 at Mautic\CoreBundle\Controller\CommonController -> delegateView ( array('viewParameters' => array('form' => object(FormView), 'currentRoute' => '/index_dev.php/s/contacts/import/new', 'mauticContent' => 'leadImport'), 'contentTemplate' => 'MauticLeadBundle:Import:new.html.php', 'passthroughVars' => array('activeLink' => '#mautic_contact_index', 'mauticContent' => 'leadImport', 'route' => '/index_dev.php/s/contacts/import/new', 'step' => '1', 'progress' => object(Progress))) )
/app/bundles/CoreBundle/Controller/CommonController.php:474 at Mautic\LeadBundle\Controller\ImportController -> newAction ( '0', '' )
/vendor/symfony/http-kernel/HttpKernel.php:135 at Mautic\CoreBundle\Controller\CommonController -> executeAction ( 'new', '0', '0', '' )
/vendor/symfony/http-kernel/HttpKernel.php:57 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), '1' )
/vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php:67 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), '1', true )
/vendor/symfony/http-kernel/Kernel.php:183 at Symfony\Component\HttpKernel\DependencyInjection\ContainerAwareHttpKernel -> handle ( object(Request), '1', true )
/app/AppKernel.php:141 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), '1', true )
/app/middlewares/CORSMiddleware.php:93 at AppKernel -> handle ( object(Request), '1', true )
/app/middlewares/CatchExceptionMiddleware.php:45 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/Dev/IpRestrictMiddleware.php:66 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/VersionCheckMiddleware.php:57 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), '1', true )
/vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), '1', true )
/vendor/stack/run/src/Stack/run.php:13 at Stack\StackedHttpKernel -> handle ( object(Request) )
/index_dev.php:33 at Stack\run ( object(StackedHttpKernel) )
Upload the CSV file that contains the items to be imported. The`

**Comment by ondrejsibl** [comment]
If I try to create a new custom field, I get this error:
Snímek obrazovky 2019-11-26 v 13 03 03

File: app/bundles/LeadBundle/Templating/Helper/AvatarHelper.php [hunk 1]
```diff
@@ -73,13 +77,13 @@ public function getAvatar(Lead $lead)
     /**
      * Get avatar path.
      *
-     * @param $absolute
+     * @param string $absolute
```
Comment (ondrejsibl) [inline app/bundles/LeadBundle/Templating/Helper/AvatarHelper.php]: This parameter can be also boolean.

**Comment by escopecz** [comment]
@ondrejsibl thanks for the bug reports! I think I fixed them all except the translation issue that I cannot replicate.

> I noticed in many places in the code you switched keys and values. Why is that?

It's described in https://docs.google.com/document/d/1lHNK8hHQtvyWlz9OtfAoLupDVb5-QunQjq-HAGlAsS4/edit#heading=h.qork18mo6yws

Ready to retest.

**Comment by hluchas** [comment]
As we talked about deprecations inside bundles. I found this in LeadBundle:
```
Targets
    Occurrences of 'deprecated' in directory /Users/lukas.drahy/dev/community-fork/app/bundles/LeadBundle
Found Occurrences  (16 usages found)
    Unclassified occurrence  (3 usages found)
        community-fork  (3 usages found)
            app/bundles/LeadBundle/Model  (3 usages found)
                LeadModel.php  (3 usages found)
                    2652 @trigger_error('getCurrentLead is deprecated and will be removed in 3.0; Use the ContactTracker::getContact instead', E_USER_DEPRECATED);
                    2669 @trigger_error('setCurrentLead is deprecated and will be removed in 3.0; Use the ContactTracker::setTrackedContact instead', E_USER_DEPRECATED);
                    2681 @trigger_error('setSystemCurrentLead is deprecated and will be removed in 3.0; Use the ContactTracker::setSystemContac instead', E_USER_DEPRECATED);
    Usage in comments  (10 usages found)
        community-fork  (10 usages found)
            app/bundles/LeadBundle/Config  (1 usage found)
                config.php  (1 usage found)
                    872 // Deprecated support for circular dependency
            app/bundles/LeadBundle/Helper  (1 usage found)
                FormFieldHelper.php  (1 usage found)
                    184 * Symfony deprecated and changed Symfony\Component\Form\Extension\Core\Type\TimezoneType::getTimezones to private
            app/bundles/LeadBundle/Model  (8 usages found)
                LeadModel.php  (7 usages found)
                    899 * @deprecated 2.0 to be removed in 3.0 - Use the Lead entity's getProfileFields() instead
                    2492 * @deprecated 2.12.0 to be removed in 3.0; use Mautic\LeadBundle\Model\DoNotContact instead
                    2532 * @deprecated 2.12.0 to be removed in 3.0; use Mautic\LeadBundle\Model\DoNotContact instead
                    2561 * @deprecated 2.12.0 to be removed in 3.0; use Mautic\LeadBundle\Model\DoNotContact instead
                    2644 * @deprecated 2.13.0 to be removed in 3.0
                    2663 * @deprecated 2.13.0 to be removed in 3.0; use ContactTracker::getContact instead
                    2689 * @deprecated 2.13.0; to be removed in 3.0. Use \Mautic\LeadBundle\Deduplicate\ContactMerger instead
                LegacyLeadModel.php  (1 usage found)
                    22 * @deprecated 2.14.0 to be removed in 3.0; Used temporarily to get around circular depdenency for LeadModel
    Usage in string constants  (3 usages found)
        community-fork  (3 usages found)
            app/bundles/LeadBundle/Model  (3 usages found)
                LeadModel.php  (3 usages found)
                    2652 @trigger_error('getCurrentLead is deprecated and will be removed in 3.0; Use the ContactTracker::getContact instead', E_USER_DEPRECATED);
                    2669 @trigger_error('setCurrentLead is deprecated and will be removed in 3.0; Use the ContactTracker::setTrackedContact instead', E_USER_DEPRECATED);
                    2681 @trigger_error('setSystemCurrentLead is deprecated and will be removed in 3.0; Use the ContactTracker::setSystemContac instead', E_USER_DEPRECATED);

```

**Comment by escopecz** [comment]
There is already an issue for deprecated LeadModel methods: https://github.com/mautic/mautic/issues/8138

That is also the last usage of the LegacyLeadModel that is deprecated in config.php: 872

FormFieldHelper.php:184 - This method is not deprecated. It's copy-pasted from `Symfony\Component\Form\Extension\Core\Type\TimezoneType::getTimezones` that is deprecated but we need it.

I don't see other deprecated code listed in your dump.

**Comment by escopecz** [comment]
I rebased it on top of latest m2-to-m3 branch to resolve conflicts.

**Comment by hluchas** [comment]
Export is failing with
```
[2019-12-06 16:24:34] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\ContextErrorException: "PHP Warning - count(): Parameter must be an array or an object that implements Countable" at /Users/lukas.drahy/dev/community-fork/app/bundles/CoreBundle/Model/IteratorExportDataModel.php line 71 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\ContextErrorException(code: 0): PHP Warning - count(): Parameter must be an array or an object that implements Countable at /Users/lukas.drahy/dev/community-fork/app/bundles/CoreBundle/Model/IteratorExportDataModel.php:71)
[stacktrace]
#0 [internal function]: Mautic\\CoreBundle\\ErrorHandler\\ErrorHandler->handleError(2, 'PHP Warning - c...', '/Users/lukas.dr...', 71, Array)
#1 /Users/lukas.drahy/dev/community-fork/app/bundles/CoreBundle/Model/IteratorExportDataModel.php(71): count(NULL)
#2 /Users/lukas.drahy/dev/community-fork/vendor/sonata-project/exporter/src/Source/IteratorSourceIterator.php(53): Mautic\\CoreBundle\\Model\\IteratorExportDataModel->next()
#3 /Users/lukas.drahy/dev/community-fork/vendor/sonata-project/exporter/src/Handler.php(43): Exporter\\Source\\IteratorSourceIterator->next()
#4 /Users/lukas.drahy/dev/community-fork/app/bundles/CoreBundle/Controller/CommonController.php(803): Exporter\\Handler->export()
#5 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-foundation/StreamedResponse.php(111): Mautic\\CoreBundle\\Controller\\CommonController->Mautic\\CoreBundle\\Controller\\{closure}()
#6 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-foundation/Response.php(383): Symfony\\Component\\HttpFoundation\\StreamedResponse->sendContent()
#7 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-kernel/EventListener/StreamedResponseListener.php(39): Symfony\\Component\\HttpFoundation\\Response->send()
#8 [internal function]: Symfony\\Component\\HttpKernel\\EventListener\\StreamedResponseListener->onKernelResponse(Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#9 /Users/lukas.drahy/dev/community-fork/vendor/symfony/event-dispatcher/Debug/WrappedListener.php(61): call_user_func(Array, Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#10 [internal function]: Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener->__invoke(Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\EventDispatcher\\ContainerAwareEventDispatcher))
#11 /Users/lukas.drahy/dev/community-fork/vendor/symfony/event-dispatcher/EventDispatcher.php(184): call_user_func(Object(Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener), Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent), 'kernel.response', Object(Symfony\\Component\\EventDispatcher\\ContainerAwareEventDispatcher))
#12 /Users/lukas.drahy/dev/community-fork/vendor/symfony/event-dispatcher/EventDispatcher.php(46): Symfony\\Component\\EventDispatcher\\EventDispatcher->doDispatch(Array, 'kernel.response', Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent))
#13 /Users/lukas.drahy/dev/community-fork/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(133): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch('kernel.response', Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent))
#14 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-kernel/HttpKernel.php(175): Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher->dispatch('kernel.response', Object(Symfony\\Component\\HttpKernel\\Event\\FilterResponseEvent))
#15 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-kernel/HttpKernel.php(157): Symfony\\Component\\HttpKernel\\HttpKernel->filterResponse(Object(Symfony\\Component\\HttpFoundation\\StreamedResponse), Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#16 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-kernel/HttpKernel.php(57): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#17 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php(67): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#18 /Users/lukas.drahy/dev/community-fork/vendor/symfony/http-kernel/Kernel.php(183): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#19 /Users/lukas.drahy/dev/community-fork/app/AppKernel.php(141): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#20 /Users/lukas.drahy/dev/community-fork/app/middlewares/CORSMiddleware.php(93): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#21 /Users/lukas.drahy/dev/community-fork/app/middlewares/CatchExceptionMiddleware.php(45): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#22 /Users/lukas.drahy/dev/community-fork/app/middlewares/Dev/IpRestrictMiddleware.php(66): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#23 /Users/lukas.drahy/dev/community-fork/app/middlewares/VersionCheckMiddleware.php(57): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#24 /Users/lukas.drahy/dev/community-fork/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#25 /Users/lukas.drahy/dev/community-fork/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#26 /Users/lukas.drahy/dev/community-fork/index_dev.php(33): Stack\
un(Object(Stack\\StackedHttpKernel))
#27 {main}
"} 
```

**Comment by escopecz** [comment]
@hluchas could you write your steps? If I go to the contact list page and select "Export" From the context menu I get a CSV of exported contacts. How to replicate your error?

**Comment by hluchas** [comment]
Export is broken on both community and this request. It's showing 

> @hluchas could you write your steps? If I go to the contact list page and select "Export" From the context menu I get a CSV of exported contacts. How to replicate your error?

I'm using the same way.

**Comment by hluchas** [comment]
It seem that this is problem of my PHP settings or dataset in DB actually.  Ignore this.

File: app/bundles/LeadBundle/Helper/TokenHelper.php [hunk 2]
```diff
@@ -39,9 +39,7 @@ public static function findLeadTokens($content, $lead, $replace = false)
         }
 
         // Search for bracket or bracket encoded
-        // @deprecated BC support for leadfield
         $tokenRegex = [
-            '/({|%7B)leadfield=(.*?)(}|%7D)/',
```
Comment (hluchas) [inline app/bundles/LeadBundle/Helper/TokenHelper.php]: Foreach below is not needed for array with one value. You could use simple variable or directly argument of `preg_match_all` function.

**Comment by escopecz** [comment]
Rebased to latest m2-to-m3 to resolve conflicts.

File: app/bundles/LeadBundle/Tests/Model/ListModelFunctionalTest.php [hunk 3]
```diff
@@ -5,206 +5,9 @@
 use Mautic\CoreBundle\Test\MauticWebTestCase;
 use Mautic\LeadBundle\Entity\LeadList;
 use Mautic\LeadBundle\Entity\LeadListRepository;
-use Monolog\Logger;
 
 class ListModelFunctionalTest extends MauticWebTestCase
 {
-    public function testSegmentCountIsCorrect()
```
[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/LeadBundle/Tests/Model/ListModelFunctionalTest.php]: Check this commit: https://github.com/escopecz/mautic/commit/1a7cbb588f6d9cef679d97372347067ce87d05a5#diff-1daccce5d6087d284e100fb78233ef26L362

The method that it's testing was deprecated and removed.

File: app/bundles/LeadBundle/Form/Type/ChangeOwnerType.php [hunk 4]
```diff
@@ -33,12 +38,13 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
-                'label'      => 'mautic.lead.batch.add_to',
-                'multiple'   => false,
-                'choices'    => $this->userModel->getOwnerListChoices(),
-                'required'   => true,
-                'label_attr' => ['class' => 'control-label'],
-                'attr'       => ['class' => 'form-control'],
+                'choices_as_values' => true,
+                'label'             => 'mautic.lead.batch.add_to',
+                'multiple'          => false,
+                'choices'           => $this->userModel->getOwnerListChoices(),
```
[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/LeadBundle/Form/Type/ChangeOwnerType.php]: Fixed in https://github.com/mautic/mautic/pull/8137/commits/c7d6e5ebdd959e0dd165dd1d9a3bea5fd4ce2feb

File: app/bundles/LeadBundle/Form/Type/CompanyMergeType.php [hunk 5]
```diff
@@ -59,7 +56,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         }
     }
 
-    public function setDefaultOptions(OptionsResolverInterface $resolver)
+    public function configureOptions(OptionsResolver $resolver)
     {
         $resolver->setOptional(
```
[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/LeadBundle/Form/Type/CompanyMergeType.php]: Updated in https://github.com/mautic/mautic/pull/8137/commits/d59a6c4af0a3a34d177bf791ef13f7d6e07b21f4

File: app/bundles/LeadBundle/Helper/TokenHelper.php (hunk 2 above)
Comment (escopecz) [reply app/bundles/LeadBundle/Helper/TokenHelper.php]: Refactored in https://github.com/mautic/mautic/pull/8137/commits/ae3da9a9d4477d8d571c2241e0beccc43aad8873

**Comment by escopecz** [comment]
Rebased again to resolved conflicts

**Pull request merged**